### DISCUSSION TITLE
Reality data and model are not aligned. Reality data is floating in air..

### DISCUSSION THREAD

**Discussion by kmurari103** [iTwin/itwinjs-core#2608] [Q&A]
Hi everyone,
I am facing this weired issue where the reality data and model are not aligned after upgrading "@bentley/imodeljs-frontend": "^2.19.18" to this new version along with all other "@bentley/..." packages. I'm using "@itwin/web-viewer-react": "^1.3.1", to create a app. These upgrade suggested me to update **await findAvailableUnattachedRealityModels(_imodel.contextId!)** as it shown deprecated in newer package. So i started using **await queryRealityData({ contextId: _imodel.contextId! });** api.

However ealier i was using @bentley/imodeljs-frontend": "^2.14.4" and **await findAvailableUnattachedRealityModels(_imodel.contextId!)**  api, it was working fine i.e Reality data and model were aligned.

Can anyone point me to what i'm missing here in new packages??

Here's snippet of how i'm doing this-
![image](https://user-images.githubusercontent.com/27068008/139434092-998ac5a0-4296-4a27-9e81-1ecc28ef5411.png)

Thanks
kmkashyap

**Comment by pmconne** [comment]
You're not missing anything; there is a bug in 2.19.18 that can cause misaligned reality data. It's been fixed and we are working on publishing new packages. I'll update shortly with a link to PR containing fix. Sorry for the temporary inconvenience.

**Comment by pmconne** [comment]
@kmkashyap The fix (#2599) has been published in @itwin/core-backend 2.19.20.

**Reply by kmurari103** [reply]
I upgraded all the packages but this issue is still there, here is my package.json file for your reference,

![image](https://user-images.githubusercontent.com/27068008/139801487-aa8a7135-21f8-4f74-9b39-91ee3046923c.png)

**Reply by KrishKash** [reply]
Hi @pmconne , sorry if you already saw the previous reply.
I updated all packages, deleted node-modules and did everything from start but the reality model is still misaligned.
Is there anything else that i'm missing?
Thanks

**Reply by pmconne** [reply]
> Is there anything else that i'm missing?

Not as far as I can tell. @AlainRobertAtBentley can you assist please?

**Reply by pmconne** [reply]
@krishna-kashyap-dev Our apologies - it turns out the fix has not yet been deployed to the backend. We'll let you know as soon as it is deployed.

**Reply by pmconne** [reply]
@krishna-kashyap-dev the fix has been deployed. Thanks for your patience -please let us know if any problems persist.

**Reply by KrishKash** [reply]
@pmconne Sorry to report that the issue didn't resolve even after updating all the packages. Here's my updated package.json, i deleted node modules and re-installed all packages but no luck. Could not figure out how it worked with older packages but started giving this issue with updated package

![image](https://user-images.githubusercontent.com/82198646/141731029-4e4e23bd-0144-4875-a465-63cf0fceae65.png)

**Reply by pmconne** [reply]
The fix was included in 2.19.20. It had to be temporarily rolled back in 2.19.21 (the version you installed). It will be reinstated in 2.19.22.

**Comment by kmurari103** [comment]
@pmconne I noticed packages (2.19.22) have been published, so i tried again - deleted node module, updated all packages to latest version, cleared browser cache but no luck. I created a [sandbox ](https://www.itwinjs.org/sandboxes/Krishna%20Kashyap/contemporary-purple-silkworm-651) with  help from one of the sample showcase  and [this snippet code](https://github.com/iTwin/itwinjs-core/discussions/2608#discussion-3656653) to depict the issue. I'm sure there is some minor thing that i may be missing here. 
Any help is appreciated. Thanks

**Reply by pmconne** [reply]
As far as I can tell the reality model is correctly geolocated, but the map is displaying flat instead of using 3d terrain, despite you instructing it to use terrain.
@Ellord207

**Reply by Rycha2021** [reply]
I am also having a similar issue. I am glad this has been raised by another user as this has been quite frustrating trying various translations.

Reality model hovering in free space regardless of any efforts i am trying to do.

Considering the delay in response (with the utmost respect), this would appear a major issue perhaps?

**Reply by aruniverse** [reply]
Just a heads up, sandbox is still on 2.19.17

**Comment by pmconne** [comment]
@Ellord207 @Josh-Schifter @kckst8 please see @kmkashyap's sandbox.
In ViewManager.onViewOpen, he configures the viewport to enable terrain. Yet the map displays as a flat plane instead of 3d terrain.
Was there not (and is there still) some bug with the initialization of the viewport in either the sandbox or the viewer? Can you explain why the change to the display style doesn't take effect and/or suggest a fix?

**Reply by pmconne** [reply]
This appears to be an issue with the iTwin viewer. @StefanRetief is investigating.

**Reply by StefanRetief** [reply]
After further investigation, it appears that the sandbox is currently looking the the "approximateTerrainHeights.json" in the incorrect location, resulting in terrain not displaying correctly. I am currently working on a fix. However, @Rycha2021 @kmkashyap, would you be able to run your build script and then run your start script and let me know if this resolves your issue? This issue would be similar to the one discussed here: https://github.com/iTwin/itwinjs-core/discussions/1355

**Reply by kmurari103** [reply]
@StefanRetief I did try these steps many times but with no luck. I even restarted my pc after deleting- /build, /node_modules folder then cleared browser cache/local storages then ran scripts- npm install, npm run build and finally npm run start.
_Note:_ I'm using "@itwin/web-viewer-react": "^1.3.4", and all "@bentley/...": "^2.19.23" packages.

**Reply by aruniverse** [reply]
Where is your viewer being served from? It's trying to fetch approximateTerrainHeights.json relative to your viewer. E.g. if your viewer is served at `/itwin/viewer/xyz` then it's going to try to fetch the above json file from  `/itwin/viewer/xyz/assets/approximateTerrainHeights.json` which probably doesn't exist. You could check the network tab and see the failing req to fetch that file. Most likely the json file can be found at the root, `assets/approximateTerrainHeights.json`

**Comment by StefanRetief** [comment]
@kmkashyap Would you happen to have any other props passed into the viewer component other than the ```_oniModelReady``` method  you've already provided? Non-memoized props can cause the viewer to re-render, which will make the viewport to reload but keep the same IModelApp instance.

**Reply by kmurari103** [reply]
Hi @StefanRetief , I do have some more props passed in viewer. Here is how viewer component looks like-
```
  return (
    <div className="viewer-container">
      <Viewer
        contextId={contextId}
        iModelId={iModelId}
        authConfig={{ config: authConfig }}
        onIModelAppInit={onIModelAppInit}
        onIModelConnected={onIModelConnected}
        viewCreatorOptions={{ viewportConfigurer: viewConfiguration }}
        uiProviders={[new SmartDeviceUiItemsProvider()]}
      />
    </div>
  );
```
onIModelAppInit contains- 

```
  const onIModelAppInit = () => {
    //console.log("onIModelAppInit " + IModelApp.authorizationClient?.hasSignedIn + "," + IModelApp.authorizationClient?.isAuthorized);
    setIsAuthorized(isAuth());

    IModelApp.authorizationClient?.onUserStateChanged.addListener(() => {      
      if (isAuth()) {
        SmartDeviceAPI.setOidcToken(IModelApp.authorizationClient!);
      }     
      setIsAuthorized(isAuth());
    });
    
    if (isAuth()) {
      SmartDeviceAPI.setOidcToken(IModelApp.authorizationClient!);
    }
   
    setIsInit(true);   
  };
```
onIModelConnected contains [this code](https://github.com/iTwin/itwinjs-core/discussions/2608#discussion-3656653)

**Reply by StefanRetief** [reply]
I suggest that you wrap the ```[new SmartDeviceUiItemsProvider]``` in a ```useMemo``` or outside of the component altogether. Typically what happens is that if you declare an array as you are currently without it, each render creates a new instance of the ```SmartDeviceUiItemsProvider```, forcing the viewer to re-render. Try it like this:

```typescript

  const uiProviders = useMemo(() => [new SmartDeviceUiItemsProvider()], []);

  return (
    <div className="viewer-container">
      <Viewer
        contextId={contextId}
        iModelId={iModelId}
        authConfig={{ config: authConfig }}
        onIModelAppInit={onIModelAppInit}
        onIModelConnected={onIModelConnected}
        viewCreatorOptions={{ viewportConfigurer: viewConfiguration }}
        uiProviders={uiProviders}
      />
    </div>
  );
```

**Reply by kmurari103** [reply]
I tried like the one you suggested above but with no luck- reality mesh still floats in the air.
I doubt if `vp.overrideDisplayStyle(displayStyle);` [in this code](https://github.com/iTwin/itwinjs-core/discussions/2608#discussion-3656653) 
 is something that is causing this issue? 
It might be conflicting with z-axis/heights or any other parameter with their default values(i might be wrong here as i dont know its internal implementation).

**Reply by StefanRetief** [reply]
@kmkashyap Are you also still having an issue with terrain not showing correctly?

**Reply by kmurari103** [reply]
Hi @StefanRetief , I guess this issue is still there as you can see in this [sandbox](https://www.itwinjs.org/sandboxes/Krishna%20Kashyap/contemporary-purple-silkworm-651). The reality mesh is still not on the ground level as expected.

**Reply by StefanRetief** [reply]
@kmkashyap I apologize for the very delayed response. I've since fixed the issues within the sandbox am now able to provide a working sandbox that no longer has a floating model nor issues loading terrain for map: https://www.itwinjs.org/sandboxes/StefanRetief/Priscilla-the-tomato-egret-757

**Reply by Rycha2021** [reply]
Promising @StefanRetief.
Is this a Global fix? 
What happens in V3 which "will" be released shortly?
Ry

**Reply by StefanRetief** [reply]
@Rycha2021 The fixes I applied were specific to the sandbox and should not normally occur outside of it, however, the code provided in the sandbox should work generally. I would like to mention that there is somewhat more performant code also available in this [sandbox](https://www.itwinjs.org/sandboxes/StefanRetief/Reality%20Data%20with%20Map). In this sandbox, instead of relying on the `onViewOpen` hooks, it will pass in a `viewState` that will render all the features you may want directly instead of trying to update the view after the view has already begun rendering. I hope that helps as well.

As for 3.0, I can't speak to those changes other than there are only minor differences in the api between 2.0 and 3.0 in regards to getting/setting reality data, but enough of a difference that it is a _breaking_ change. There will be a [codemod](https://github.com/iTwin/codemods) that should make the experience of upgrading to 3.0 as painless as possible. As soon as I can, I will also update this thread with a 3.0 version of the sandbox mentioned above.

**Reply by kmurari103** [reply]
Thanks @StefanRetief , that worked perfectly fine.
Also thanks to everyone involved in addressing this issue. Really appreciate all. 
Cheers 😃

**Reply by pmconne** [reply]
@kmkashyap @StefanRetief's workaround is also a bug. The actual fix is in #2949. This fix will cause the "workaround" to stop working (i.e., it will correctly **not** display terrain).
